x/log_writer.go

```diff
@@ -236,6 +306,109 @@ func (l *LogWriter) open() error {
 		return info.Size()
 	}
 
+	openNew := func() error {
+		f, err := os.OpenFile(l.FilePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
+		if err != nil {
+			return err
+		}
+		l.file = f
+		l.writer = bufio.NewWriterSize(l.file, bufferSize)
+
+		if l.EncryptionKey != nil {
+			iv := make([]byte, 16)
+			if _, err := rand.Read(iv); err != nil { // cve fix is here
+				return err
+			}
+			lengthInput := make([]byte, 4)
+			binary.BigEndian.PutUint32(lengthInput, uint32(len(VerificationText))) // header has 16+4 bytes now
+
+			bytes, err := encrypt(l.EncryptionKey, iv, []byte(VerificationText))
+			cipher := append(append(iv, lengthInput...), bytes...)
+			if _, err = l.writer.Write(cipher); err != nil {
+				return err
+			}
+		}
+		l.size = size()
+		return nil
+	}
+
+	info, err := os.Stat(l.FilePath)
+	if err != nil { // if any error try to open new log file itself
+		return openNew()
+	}
+
+	// encryption is enabled and file is corrupted as not able to read the IV
+	if l.EncryptionKey != nil && info.Size() < 12 {
+		return openNew()
+	}
+
+	f, err := os.OpenFile(l.FilePath, os.O_APPEND|os.O_RDWR, os.ModePerm)
+	if err != nil {
+		return openNew()
+	}
+
+	l.file = f
+	if l.EncryptionKey != nil {
+		// initialize byte slice for iv
+		iv := make([]byte, 16)
+		// If not able to read the iv, then this file might be corrupted.
+		// open the new file in that case
+		if _, err = l.file.ReadAt(iv, 0); err != nil {
+			_ = l.file.Close()
+			return openNew()
+		}
+		text := make([]byte, 4+len(VerificationText)) // size=4+11
+		// veritification text starts at offset 20, however in old audit logs it starts at 16
+		// we may encounter an old audit log and need to decrypt it
+		if _, err := f.ReadAt(text, 16); err != nil {
+			_ = f.Close()
+			return openNew()
+		}
+		t, err := decrypt(l.EncryptionKey, iv, text[4:])
+		if err != nil {
+			_ = f.Close()
+			return openNew()
+		} else if string(t) != VerificationText {
+			// try old decrypt method in case we have old audit log
+			t = make([]byte, 11)
+			copy(t, text[:11]) // decrypt_deprecated destroys bytes, need copy here
+			t, err = decrypt_deprecated(l.EncryptionKey, l.baseIv, t)
+			if err != nil || string(t) != VerificationText_Deprecated {
+				// file corrupted or wrong key
+				_ = f.Close()
+				return openNew()
+			} else {
+				// rotate old audit log
+				if err := l.rotate(); err != nil {
+					return err
+				}
+				return nil
+			}
+		}
+	}
+
+	l.writer = bufio.NewWriterSize(l.file, bufferSize)
+	l.size = size()
+	return nil
+}
+
+func (l *LogWriter) open_deprecated() error {
```

Comment:
Have we added these functions?

Reply:
this is the code that existed there previously - I just appended _deprecated to the function name.  If you feel it is not necessary to keep this function in the code I can remove it

Reply:
we didn't add the replacement code here then? and I think we can remove the code that is not used, we can always find in the history.

Reply:
sounds reasonable to me - I remove the unused code and kept only the functions necessary to decrypt old audit logs